src/api/ledgerManager.js

```diff
@@ -54,6 +54,42 @@ export class LedgerManager {
     return diffLedger(remoteLedger, this._ledger);
   }
 
+  /**
+   * Persists the local (in-memory) Ledger to the ledger storage. Reloads the
+   * remote ledger from storage right before persisting it to minimize the
+   * possibility of overwriting remote changes that were not synced to the local
+   * ledger and ensure consistency of the ledger events (e.g. no double spends).
+   *
+   * A race condition is present in this function: if client A runs reloadLedger
+   * and then client B writes to the remote ledger before client A finishes writing,
+   * then the changes to the ledger that client B made would be overwritten by the
+   * changes from client A. The correctness and consistency of the ledger will still
+   * be maintained, its just that client B might experience data loss of whatever
+   * events they were trying to sync. To detect if this has occurred, we reload
+   * the ledger again after writing the data to ensure the local changes were
+   * not overwritten. If they were, we can show an error message to client B with
+   * a list of changes that failed to sync.
+   */
+  async persist(): Promise<ReloadResult> {
+    // START RACE CONDITION
+    let reloadRes = await this.reloadLedger();
+    if (reloadRes.error) {
+      return reloadRes;
+    }
+    await this._storage.write(this._ledger);
+    // END RACE CONDITION
+
+    // Reload ledger again to ensure all the changes were persisted into storage
+    reloadRes = await this.reloadLedger();
+    if (reloadRes.localChanges.length !== 0) {
+      return {
+        ...reloadRes,
+        error: "Some local changes have not been persisted",
+      };
+    }
```

Comment:
if `reloadRes` has an error on this second load, then we are silently discarding the error message. I think we should check for error, explicitly return with the error if there was one, and only then try to see whether we had a problem with localChanges.